wpical/src/main/native/cpp/fieldcalibration.cpp

```diff
@@ -0,0 +1,600 @@
+
+  Eigen::Matrix<double, 4, 4> camera_to_tag{
+      {r_mat.at<double>(0, 0), r_mat.at<double>(0, 1), r_mat.at<double>(0, 2),
+       t_vec.at(0)},
+      {r_mat.at<double>(1, 0), r_mat.at<double>(1, 1), r_mat.at<double>(1, 2),
+       t_vec.at(1)},
+      {r_mat.at<double>(2, 0), r_mat.at<double>(2, 1), r_mat.at<double>(2, 2),
+       t_vec.at(2)},
+      {0.0, 0.0, 0.0, 1.0}};
+
+  return camera_to_tag;
+}
+
+inline void draw_tag_cube(cv::Mat& frame,
+                          Eigen::Matrix<double, 4, 4> camera_to_tag,
+                          const Eigen::Matrix<double, 3, 3>& camera_matrix,
+                          const Eigen::Matrix<double, 8, 1>& camera_distortion,
+                          double tag_size) {
+  cv::Mat camera_matrix_cv;
+  cv::Mat camera_distortion_cv;
+
+  cv::eigen2cv(camera_matrix, camera_matrix_cv);
+  cv::eigen2cv(camera_distortion, camera_distortion_cv);
+
+  std::vector<cv::Point3f> points_3d_box_base = {
+      cv::Point3f(-tag_size / 2.0, tag_size / 2.0, 0.0),
+      cv::Point3f(tag_size / 2.0, tag_size / 2.0, 0.0),
+      cv::Point3f(tag_size / 2.0, -tag_size / 2.0, 0.0),
+      cv::Point3f(-tag_size / 2.0, -tag_size / 2.0, 0.0)};
+
+  std::vector<cv::Point3f> points_3d_box_top = {
+      cv::Point3f(-tag_size / 2.0, tag_size / 2.0, -tag_size),
+      cv::Point3f(tag_size / 2.0, tag_size / 2.0, -tag_size),
+      cv::Point3f(tag_size / 2.0, -tag_size / 2.0, -tag_size),
+      cv::Point3f(-tag_size / 2.0, -tag_size / 2.0, -tag_size)};
+
+  std::vector<cv::Point2f> points_2d_box_base = {
+      cv::Point2f(0.0, 0.0), cv::Point2f(0.0, 0.0), cv::Point2f(0.0, 0.0),
+      cv::Point2f(0.0, 0.0)};
+
+  std::vector<cv::Point2f> points_2d_box_top = {
+      cv::Point2f(0.0, 0.0), cv::Point2f(0.0, 0.0), cv::Point2f(0.0, 0.0),
+      cv::Point2f(0.0, 0.0)};
+
+  Eigen::Matrix<double, 3, 3> r_vec = camera_to_tag.block<3, 3>(0, 0);
+  Eigen::Matrix<double, 3, 1> t_vec = camera_to_tag.block<3, 1>(0, 3);
+
+  cv::Mat r_vec_cv;
+  cv::Mat t_vec_cv;
+
+  cv::eigen2cv(r_vec, r_vec_cv);
+  cv::eigen2cv(t_vec, t_vec_cv);
+
+  cv::projectPoints(points_3d_box_base, r_vec_cv, t_vec_cv, camera_matrix_cv,
+                    camera_distortion_cv, points_2d_box_base);
+  cv::projectPoints(points_3d_box_top, r_vec_cv, t_vec_cv, camera_matrix_cv,
+                    camera_distortion_cv, points_2d_box_top);
+
+  for (int i = 0; i < 4; i++) {
+    cv::Point2f& point_base = points_2d_box_base.at(i);
+    cv::Point2f& point_top = points_2d_box_top.at(i);
+
+    cv::line(frame, point_base, point_top, cv::Scalar(0, 255, 255), 5);
+
+    int i_next = (i + 1) % 4;
+    cv::Point2f& point_base_next = points_2d_box_base.at(i_next);
+    cv::Point2f& point_top_next = points_2d_box_top.at(i_next);
+
+    cv::line(frame, point_base, point_base_next, cv::Scalar(0, 255, 255), 5);
+    cv::line(frame, point_top, point_top_next, cv::Scalar(0, 255, 255), 5);
+  }
+}
+
+inline bool process_video_file(
+    apriltag_detector_t* tag_detector,
```

Comment:
Nit: Now that we are in allwpilib, it should be easy to use the allwpilib AprilTag wrappers. Their API is a whole lot more sane then the umich C api, LOL. Worth considering for future.

Reply:
Yeah, it would be good to use the AprilTag wrappers, but that would require wpimath and we'd have conflicting Eigen versions, which our build can't handle (yet.)

Reply:
And that's also why we can't use AprilTagFieldLayout yet right?

Reply:
Yep.